discovery-provider/src/utils/ipfs_lib.py

```diff
@@ -40,178 +40,137 @@ def __init__(
-                    "IPFSCLIENT | get_metadata_from_gateway \
-                        \nfailed to fetch metadata from user replica gateways"
-                )
+                for future in asyncio.as_completed(futures, timeout=5):
+                    try:
+                        future_result = await future
+                    except asyncio.CancelledError:
+                        pass  # swallow canceled requests
 
-        logger.warning(
-            f"IPFSCLIENT | get_metadata_from_gateway, \
-                \ncombined addresses: {gateway_endpoints}, \
-                \ncnode_endpoints: {self._cnode_endpoints}"
-        )
-
-        query_urls = [f"{addr}/ipfs/{multihash}" for addr in gateway_endpoints]
-        data = self.query_ipfs_metadata_json(query_urls, default_metadata_fields)
-        if data is None:
-            raise Exception(
-                f"IPFSCLIENT | Failed to retrieve CID {multihash} from gateway"
-            )
-        gateway_metadata_json = data
-        return gateway_metadata_json
+                    if not future_result:
+                        continue
 
-    def update_cnode_urls(self, cnode_endpoints):
-        if len(cnode_endpoints):
-            logger.info(
-                f"IPFSCLIENT | update_cnode_urls with endpoints {cnode_endpoints}"
-            )
-            self._cnode_endpoints = cnode_endpoints
+                    cid, metadata_json = future_result
 
+                    metadata_format = (
+                        track_metadata_format
+                        if cid_type[cid] == "track"
+                        else user_metadata_format
+                    )
+
+                    formatted_json = self._get_metadata_from_json(
+                        metadata_format, metadata_json
+                    )
 
-def construct_image_dir_gateway_url(address, CID):
-    """Construct the gateway url for an image directory.
+                    if formatted_json != metadata_format:
+                        cid_metadata[cid] = formatted_json
 
-    Args:
-        args.address - base url of gateway
-        args.CID - CID of the image directory
-    """
-    if not address:
-        return None
+                        if len(fetched_cids) + len(cid_metadata) == len(
+                            cids_txhash_set
+                        ):
+                            break  # fetched all metadata
 
-    return urljoin(address, f"/ipfs/{CID}/original.jpg")
+                        for other_future in cid_futures_map[cid]:
+                            if other_future == future:
+                                continue
+                            other_future.cancel()  # cancel other pending requests
+
+            except asyncio.TimeoutError:
+                logger.info(
+                    "IPFSCLIENT | fetch_metadata_from_gateway_endpoints TimeoutError"
+                )
+            except Exception as e:
+                logger.info("IPFSCLIENT | Error in fetch cid metadata")
```

Comment:
Let's log the exception here

Reply:
The exception gets raised and logged here https://github.com/AudiusProject/audius-protocol/blob/14af524fd6de0a8bcac95424f71ec57095ea9df9/discovery-provider/src/tasks/index.py#L692